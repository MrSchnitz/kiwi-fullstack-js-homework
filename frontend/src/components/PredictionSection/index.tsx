import React from "react";
import styled from "styled-components";
import Prediction from "./Prediction";
import { WordType } from "../../models/Word";
import { Loading } from "@kiwicom/orbit-components/lib";

interface PredictionSectionProps {
  predictions: WordType[];
  predictionsLoading?: boolean;
  onSelectPrediction: (word: WordType) => void;
}

const PredictionSection: React.FC<PredictionSectionProps> = ({
  predictions,
  predictionsLoading,
  onSelectPrediction,
}: PredictionSectionProps) => {
  return (
    <PredictionSectionContent
      loading={!!predictionsLoading}
      show={predictions.length > 0 || !!predictionsLoading}
    >
      {predictionsLoading ? (
        <Loading type={"inlineLoader"} />
      ) : (
        predictions.length > 0 &&
        predictions.map((prediction: WordType) => (
          <Prediction
            key={prediction.id}
            text={prediction.word}
            onClick={() => onSelectPrediction(prediction)}
          />
        ))
      )}
    </PredictionSectionContent>
  );
};

interface PredictionContentInterface {
  loading: boolean;
  show: boolean;
}

const PredictionSectionContent = styled.div<PredictionContentInterface>`
  height: ${(props) => (props.show ? "7.5rem" : 0)};
  width: 100%;
  margin-bottom: 0.3rem;
  overflow-x: auto;
  background-color: #008f7b;
  display: flex;
  justify-content: ${(props) => (props.loading ? "center" : "flex-start")};
  align-items: center;

  transition: 0.5s height;

  &::-webkit-scrollbar {
    height: 0.8rem;
    transition: 0.3s all;
  }
  &::-webkit-scrollbar-track {
    background: #fff;
  }
  &::-webkit-scrollbar-thumb {
    background: #008f7b;
    border: 1px solid #fff;
    transition: 1s all;
  }
  &::-webkit-scrollbar-thumb:hover {
    background: #008f7b8c;
  }
`;

export default PredictionSection;
