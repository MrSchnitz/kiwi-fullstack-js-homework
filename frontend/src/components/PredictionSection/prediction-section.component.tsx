import React from "react";
import Prediction from "../Prediction/prediction.component";
import { WordType } from "../../models/Word";
import { Loading } from "@kiwicom/orbit-components/lib";
import { PredictionSectionContent } from "./prediction-section.styles";

export interface PredictionSectionProps {
  predictions: WordType[];
  predictionsLoading?: boolean;
  onSelectPrediction: (word: WordType) => void;
}

const PredictionSection: React.FC<PredictionSectionProps> = ({
  predictions,
  predictionsLoading,
  onSelectPrediction,
}: PredictionSectionProps) => {
  return (
    <PredictionSectionContent
      isLoading={!!predictionsLoading}
      show={predictions.length > 0 || !!predictionsLoading}
    >
      {predictionsLoading ? (
        <Loading type={"inlineLoader"} />
      ) : (
        predictions.length > 0 &&
        predictions.map((prediction: WordType) => (
          <Prediction
            key={prediction.id}
            text={prediction.word}
            onClick={() => onSelectPrediction(prediction)}
          />
        ))
      )}
    </PredictionSectionContent>
  );
};

export default PredictionSection;
