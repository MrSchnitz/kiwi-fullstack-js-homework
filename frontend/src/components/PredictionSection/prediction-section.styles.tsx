import styled from "styled-components";

interface PredictionContentInterface {
  isLoading: boolean;
  show: boolean;
}

export const PredictionSectionContent = styled.div<PredictionContentInterface>`
  height: ${(props) => (props.show ? "7.5rem" : 0)};
  width: 100%;
  margin-bottom: 0.3rem;
  overflow-x: auto;
  background-color: #008f7b;
  display: flex;
  justify-content: ${(props) => (props.isLoading ? "center" : "flex-start")};
  align-items: center;

  transition: 0.5s height;

  &::-webkit-scrollbar {
    height: 0.8rem;
    transition: 0.3s all;
  }
  &::-webkit-scrollbar-track {
    background: #fff;
  }
  &::-webkit-scrollbar-thumb {
    background: #008f7b;
    border: 1px solid #fff;
    transition: 1s all;
  }
  &::-webkit-scrollbar-thumb:hover {
    background: #008f7b8c;
  }
`;

PredictionSectionContent.displayName = "PredictionSectionContent";
