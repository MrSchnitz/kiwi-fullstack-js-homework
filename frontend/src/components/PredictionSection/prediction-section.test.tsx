import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import PredictionSection, {
  PredictionSectionProps,
} from "../PredictionSection/prediction-section.component";

describe("PredictionSection component", () => {
  let wrapper: ShallowWrapper;
  let mockOnSelectPrediction: any;
  let mockProps: PredictionSectionProps;

  beforeEach(() => {
    mockOnSelectPrediction = jest.fn();

    mockProps = {
      onSelectPrediction: mockOnSelectPrediction,
      predictions: [
        { id: "0", word: "home" },
        { id: "1", word: "good" },
      ],
      predictionsLoading: false,
    };

    wrapper = shallow(<PredictionSection {...mockProps} />);
  });

  it("should render PredictionSection component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should not display content", () => {
    mockProps.predictions = [];
    wrapper = shallow(<PredictionSection {...mockProps} />);

    expect(wrapper.find("PredictionSectionContent").prop("loading")).toBe(
      false
    );
    expect(wrapper.find("PredictionSectionContent").prop("show")).toBe(false);
  });

  it("should display Prediction components", () => {
    expect(wrapper.find("PredictionSectionContent").children().length).toBe(2);
  });

  it("should return prediction when Prediction child component is clicked", () => {
    wrapper.find("PredictionSectionContent").childAt(0).simulate("click");

    expect(mockOnSelectPrediction).toHaveBeenCalled();
    expect(mockOnSelectPrediction.mock.calls[0][0]).toBe(
      mockProps.predictions[0]
    );
  });
});
