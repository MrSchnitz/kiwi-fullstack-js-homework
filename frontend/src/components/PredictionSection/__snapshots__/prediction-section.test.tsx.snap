// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PredictionSection component should render PredictionSection component 1`] = `ShallowWrapper {}`;
